jerry-core/api/jerry.c

```diff
@@ -201,23 +192,39 @@ jerry_cleanup (void)
 
   jmem_finalize ();
   jerry_make_api_unavailable ();
-
-  if (JERRY_CONTEXT (user_context_deinit_cb))
-  {
-    JERRY_CONTEXT (user_context_deinit_cb) (JERRY_CONTEXT (user_context_p));
-  }
 } /* jerry_cleanup */
 
 /**
- * Retrieve user context.
+ * Retrieve a context data item, or create a new one.
+ *
+ * @param manager_p pointer to the manager whose context data item should be returned.
  *
- * @return the user-provided context-specific pointer
+ * @return a pointer to the user-provided context-specific data item for the given manager, creating such a pointer if
+ * none was found.
  */
 void *
-jerry_get_user_context (void)
+jerry_get_context_data (const jerry_context_data_manager_t *manager_p)
 {
-  return JERRY_CONTEXT (user_context_p);
-} /* jerry_get_user_context */
+  void *ret = NULL;
+  jerry_context_data_header_t *item_p;
+
+  for (item_p = JERRY_CONTEXT (context_data_p); item_p != NULL; item_p = item_p->next_p)
+  {
+    if (item_p->manager_p == manager_p)
+    {
+      return JERRY_CONTEXT_DATA_HEADER_USER_DATA (item_p);
+    }
+  }
+
+  item_p = jmem_heap_alloc_block (sizeof (jerry_context_data_header_t) + manager_p->bytes_needed);
```

Comment:
Yet another alternative is to re-introduce the *a priori* aspect and initialize each context with a fixed number of extra slots:
```c
void jerry_init (jerry_init_flag_t flags, size_t context_slot_count);
```
Then we could have an array of
```c
struct {
  jerry_context_manager_t *manager_p;
  void *user_data_p
};
```
which we could allocate contiguously with the rest of the context if we stick it to the end of the context, thus saving a further word.

This is kinda limiting though because the value of `context_slot_count` will need to be known ahead of time, and there might be some temptation to make that value large in anticipation of future complex code.